### DISCUSSION TITLE
VDC GUI

### DISCUSSION THREAD

**Discussion by johnpalermo** [c128lib/.github#12] [Team Posts]
Cia Raffaele,

Just saw your new branch for a VDC GUI. Very cool. Let me know when finished. I would like to try it out for fun. Is it working good in VICE?

**Comment by intoinside** [comment]
Hi John, you can already try some macros working. You can find it in lib/vdc-gui-global.asm
They works correctly but i want to optimize as much as possible before release them officially.
They also works good in VICE.
If you want to try, let me know your feedback (new macros will come in the next future)

**Reply by johnpalermo** [reply]
Good first version. I was playing with some code and it looks really good. I tried the progress bar also and sometimes there's a blank character showing, like this:

![Screenshot 2023-08-31 125136](https://github.com/c128lib/.github/assets/85082089/e4ab6256-5eeb-44fb-9c65-32abfa37ba33)

Here's the code for that:
```nasm
#define VDC_POSITIONATTRXY
#define VDC_CREATEWINDOW

c128lib_BasicUpstart128(start)

text1:
    .text "first window"
text2:
    .text "second window"
text3:
    .text "third window"

start:

    c128lib_CreateWindow(1,1,79,24)
    c128lib_CreateWindowWithTitle(3,3,40,12,text1,12)
    c128lib_CreateWindowWithTitleColor(7,7,40,10,text2,13,c128lib.Vdc.VDC_LIGHT_RED)
    c128lib_CreateWindowWithTitleColor(15,10,44,12,text3,12,c128lib.Vdc.VDC_DARK_YELLOW)
    c128lib_ProgressBar(18,19,40,10,3)
    
#import "./common/lib/common-global.asm"
#import "./framework/lib/vdc-gui-global.asm"
#import "./chipset/lib/vdc-global.asm"       
```
Also, I passed 10 steps and position 3 to the macro, but it looks like 11 steps showing and 4 steps completed with a blank at the beginning. 

Sometimes the progress bar looks like this:
![Screenshot 2023-08-31 132426](https://github.com/c128lib/.github/assets/85082089/f7ab1e3c-cc5b-42cd-b639-300da0bb37fd)

I only changed the last 2 lines. Screen shows 1 more step and position than passed to the macro.

```nasm
    c128lib_CreateWindowWithTitleColor(15,10,40,12,text3,12,c128lib.Vdc.VDC_DARK_YELLOW)
    c128lib_ProgressBar(18,19,36,10,8)
```

I like the windows. I tried overlapping windows just for fun. Obviously, they appear transparent. A nice option would be to set them to appear opaque.

**Comment by intoinside** [comment]
Yes i think opaque window is a better solution/option, i'll set an issue about it. Progressbar is still WIP, and it's very buggy. Also there are some contraints about parameters so it's better to rewrite it
Thanks for your test, you gave me some useful hints!

**Comment by intoinside** [comment]
@johnpalermo i made a huge refactor for window creation, now it should looks better and there are some useful parameter to use. You can see some calling example on inline documentation in vdc-gui-global.asm
Progress bar is still a mess but it's the next work item

**Reply by johnpalermo** [reply]
Cool. I will check Friday as I'll be away from home until Thursday night. 😊

**Reply by johnpalermo** [reply]
Looks really good. The huge refactor is much better. Some playing around:
![Screenshot 2023-09-10 160703](https://github.com/c128lib/.github/assets/85082089/6f9ae439-6312-4638-8f77-cb2c411440bd)

What do you think about making the title background opaque for the full width of the window, like the first window below.

![Screenshot 2023-09-10 160703 - Copy](https://github.com/c128lib/.github/assets/85082089/de62ae46-c12d-482f-8298-9659373218eb)

**Comment by intoinside** [comment]
Yes, i think opaque background for title is a good idea. I've created a button macro issue but it would be useful only if user can interact with gui. Now gui it's only painted on screen and i need to study how to enable interaction (probably in another release after 0.2.0).
First release will be only paint.
